musicbrainz/csv/get_genre.py

```diff
@@ -0,0 +1,64 @@
+"""
+Script for retrieving the "genre" dumps since they are not available online.
+"""
+
+import time
+import requests
+from requests.exceptions import HTTPError
+import pandas as pd
+from bs4 import BeautifulSoup
+
+URL = "https://musicbrainz.org/ws/2/genre/all"
+PARAMS = dict(fmt="json", limit="50")
+MAX_REQUEST_RETRIES = 3
+
+data = []
+max_records = requests.get(url=URL, params=PARAMS, timeout=500).json()["genre-count"]
+
+for i in range(0, max_records, 50):
+    PARAMS["offset"] = i
+    for j in range(MAX_REQUEST_RETRIES):
```

Comment:
We were having this discussion in the other thread, I think, but I don't think the result of that made its way here yet. 

Reviewing the (API documentation for MusicBrainz)[https://musicbrainz.org/doc/MusicBrainz_API/Rate_Limiting]:
1. They want an informative `User-Agent` header so they can contact us if something is going wrong. 
2. Read the information about rate limiting based on source IP. Have we constructed this to be within those limits?
3. If they reach the global limit, how does this script respond?

Reply:
> They want an informative User-Agent header so they can contact us if something is going wrong.
I still don't understand what is an User-Agent header.
> Read the information about rate limiting based on source IP. Have we constructed this to be within those limits?
They said the maximum rate is 50 request per second, I wait for 0.02s each for loop iteration. In case an HTTPError occurs, I wait for an longer time.

> Source IP address
> Unless you have agreed otherwise with MusicBrainz, the rule is as follows:
> 
> The rate at which your IP address is making requests is measured. If that rate is too high, all your requests will be declined (http 503) until the rate drops again. Currently that rate is (on average) 1 request per second.
> 
> For example: if your requests are coming in at 4 requests per second, we don't honour 25% of them and decline the other 75% - we decline 100% of them, until the rate drops to 1 per second or lower.

This is on the website. What does it mean to have agreed with MusicBrainz?

> If they reach the global limit, how does this script respond?
As long as they returns a HTTPError, the script drops the rate by waiting a longer time for each iteration.

Reply:
> I still don't understand what is an User-Agent header.

Happy to help! What do you know about it and what is confusing?

> They said the maximum rate is 50 request per second

This is incorrect. The part you quote does not say 50 requests per second. It says "(on average) 1 request per second". 

> What does it mean to have agreed with MusicBrainz?

It says "unless you have agreed with MusicBrainz". We don't have any agreement that I know of with MusicBrainz, so I think we are covered by their generic rules: "Rate is (on average) 1 request per second".

Reply:
> As long as they returns a HTTPError, the script drops the rate by waiting a longer time for each iteration.

Ok

Reply:
> I still don't understand what is an User-Agent header.

What is it's format and how do we pass it to the MusicBrainz API?

Reply:
The API documentation page gives a nice suggested format for the User-Agent header:

![image](https://github.com/user-attachments/assets/0a88ae8c-6c93-4cc6-86e5-75173d694c11)

See what you can find in the requests package documentation about how to pass a request header as part of your request.

Reply:
Headers look great!